File: 2022/day10/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Solution {

    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get(System.getProperty("user.dir"), "..", "input.txt");
        List<String> instructions = Files.readAllLines(inputPath);

        System.out.println("Part 1: " + part1(instructions));
        System.out.println("Part 2:");
        System.out.println(part2(instructions));
    }

    static int part1(List<String> instructions) {
        Set<Integer> targetCycles = new HashSet<>(Arrays.asList(20, 60, 100, 140, 180, 220));
        int total = 0;
        int x = 1;
        int cycle = 0;

        for (String line : instructions) {
            if (line.equals("noop")) {
                cycle++;
                if (targetCycles.contains(cycle)) {
                    total += cycle * x;
                }
            } else { // addx V
                int v = Integer.parseInt(line.split(" ")[1]);
                cycle++;
                if (targetCycles.contains(cycle)) {
                    total += cycle * x;
                }
                cycle++;
                if (targetCycles.contains(cycle)) {
                    total += cycle * x;
                }
                x += v;
            }
        }

        return total;
    }

    static String part2(List<String> instructions) {
        StringBuilder screen = new StringBuilder();
        StringBuilder row = new StringBuilder();
        int x = 1;
        int cycle = 0;

        for (String line : instructions) {
            if (line.equals("noop")) {
                cycle++;
                int pos = (cycle - 1) % 40;
                row.append(Math.abs(pos - x) <= 1 ? '#' : '.');
                if (cycle % 40 == 0) {
                    screen.append(row).append('\n');
                    row = new StringBuilder();
                }
            } else { // addx V
                int v = Integer.parseInt(line.split(" ")[1]);

                // First cycle
                cycle++;
                int pos = (cycle - 1) % 40;
                row.append(Math.abs(pos - x) <= 1 ? '#' : '.');
                if (cycle % 40 == 0) {
                    screen.append(row).append('\n');
                    row = new StringBuilder();
                }

                // Second cycle
                cycle++;
                pos = (cycle - 1) % 40;
                row.append(Math.abs(pos - x) <= 1 ? '#' : '.');
                if (cycle % 40 == 0) {
                    screen.append(row).append('\n');
                    row = new StringBuilder();
                }

                x += v;
            }
        }

        // Remove trailing newline if present
        if (screen.length() > 0 && screen.charAt(screen.length() - 1) == '\n') {
            screen.setLength(screen.length() - 1);
        }

        return screen.toString();
    }
}

File: 2022/day10/rust/solution.rs
use std::fs;
use std::path::Path;

fn simulate_cpu(instructions: &[&str]) -> Vec<(i32, i32)> {
    let mut x: i32 = 1;
    let mut cycle: i32 = 0;
    let mut cycles = Vec::new();

    for line in instructions {
        if *line == "noop" {
            cycle += 1;
            cycles.push((cycle, x));
        } else {
            // addx V
            let v: i32 = line.split_whitespace().nth(1).unwrap().parse().unwrap();
            cycle += 1;
            cycles.push((cycle, x));
            cycle += 1;
            cycles.push((cycle, x));
            x += v;
        }
    }

    cycles
}

fn part1(instructions: &[&str]) -> i32 {
    let target_cycles = [20, 60, 100, 140, 180, 220];
    let mut total = 0;

    for (cycle, x) in simulate_cpu(instructions) {
        if target_cycles.contains(&cycle) {
            total += cycle * x;
        }
    }

    total
}

fn part2(instructions: &[&str]) -> String {
    let mut screen = Vec::new();
    let mut row = String::new();

    for (cycle, x) in simulate_cpu(instructions) {
        let pos = (cycle - 1) % 40;
        if (pos - x).abs() <= 1 {
            row.push('#');
        } else {
            row.push('.');
        }

        if cycle % 40 == 0 {
            screen.push(row.clone());
            row.clear();
        }
    }

    screen.join("\n")
}

fn main() {
    let exe_path = std::env::current_exe().unwrap();
    let exe_dir = exe_path.parent().unwrap();
    let input_path = exe_dir.join("../input.txt");

    // Fallback to relative path if exe-relative doesn't exist
    let input_path = if input_path.exists() {
        input_path
    } else {
        Path::new("../input.txt").to_path_buf()
    };

    let content = fs::read_to_string(&input_path).expect("Failed to read input file");
    let instructions: Vec<&str> = content.trim().split('\n').collect();

    println!("Part 1: {}", part1(&instructions));
    println!("Part 2:");
    println!("{}", part2(&instructions));
}
